### ISSUE TITLE
[Feature] Support for 1080p Enhanced Bitrate

### ISSUE THREAD

**Issue by tadly** [anxdpanic/plugin.video.youtube#505]
### What Problem Does This Solve? 
Google (apparently) has been rolling out a [Enhanced Bitrate version of the 1080p stream](https://www.theverge.com/2023/4/10/23677141/youtube-premium-subscribers-higher-quality-video-1080p).
This is only for premium users though.

Would it be possible for the addon to support this as well?
With the videos I'm watching (mostly gaming related content), the default 1080p stream oftentimes looks so bad you
can't make out anything due to the insane compression.

**Comment by MoojMidge** [comment]
Do you have YouTube Premium?

It may be possible to do but it would need you to log in on a browser, extract your cookies, and add a new feature in the plugin to use the extracted cookie data.

However because the plugin doesn't use the browser client details for fetching streams, the browser cookies may not be directly compatible, and I have no way to debug it without getting YouTube Premium myself and no intention to get it either.

**Comment by tadly** [comment]
I do have Premium, yes.
So currently, the addon get's the streams without a login-session? Didn't know that :)

Is it using youtube-dl for this?
If so, maybe that's where this feature would have to be added.

**Comment by MoojMidge** [comment]
No thsi plugin doesn't use youtube-dl, but I think that is how youtube-dl handles this already.

**Comment by drelephant** [comment]
I've noticed that `yt-dlp` can download the high bitrate stream even without having Youtube Premium.

Not sure if that means that this plugin should be able to access it as well.

**Comment by MoojMidge** [comment]
The only premium streams I have seen are only available for iOS clients.

yt-dlp mimics a variety of different clients, including Android, iOS, and a web browser, to get a number of different quality streams.

This plugin primarily mimics an Android client, and I have yet to see any premium streams on Android. It may require having Youtube Premium and sharing your browser cookie data to make any such streams accessible, if they exist.

Any streams that are available to the client this plugin mimics can generally also be used by the plugin, although some streams may be obfuscated or throttled.

Theoretically the plugin can mimic an iOS client and play any premium streams that are available. Not actually theoretical - I have tried it and it does work, but unlike yt-dlp, the plugin only tries to mimic other clients if it can't access a video at all.

So it is possible but requires a few changes to make it work.

**Comment by tadly** [comment]
thanks for that write-up :)
I just want to add that I have seen the premium stream on youtube web as well (so not only limited to iOS) although I haven't seen one in some time (just clicked through a few and none have that option available).
Maybe they removed it in the web-version again :/

>I have tried it

that means you adapted the code to simulate iOS and you were given the option of a premium stream?
How much effort would it be do make that a setting for the addon? A "Preferred client type" option with Android, iOS, Web availble to choose from

**Comment by MoojMidge** [comment]
> that means you adapted the code to simulate iOS and you were given the option of a premium stream?
> How much effort would it be do make that a setting for the addon? A "Preferred client type" option with Android, iOS, Web availble to choose from

Web clients are not currently usable in this plugin - the associated streams are obfuscated and throttled and the code necessary to enable their use requires significant work.

Most of the code required to use an iOS client already exists in v7.0.2+ of the plugin, and with some additional minor changes when an iOS client is used the premium streams are extracted and can be played. However the user experience is not as good, because the type of streams are more limited on iOS compared to Android.

So while you may get a 1080p premium stream with the iOS client, you may not get a better quality 2k or 4k stream that is available using the Android client, because the plugin currently does not have the ability to use multiple client requests at the same time i.e. it can't use both the Android client and the iOS client to select the best streams from both.

The functionality to prefer different clients also exists in v7.0.2+, but the iOS client is not currently enabled.

**Comment by tadly** [comment]
very interesting... I only stream 1080p anyways as with higher streams I experience buffering more frequently.
Also old hardware struggles with 60fps playback on those :')

TL;DR
1080p high bitrate is exactly what I would want
